### PR TITLE
feat: dirty components

### PR DESCRIPTION
**Pull request by belohlavek**
Introduces a DataObject decorator that allows to easily create proxy properties that can set the `isDirty` field of a component to true.

### CODE REVIEW

File: src/core/DataObject.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  // @internal
+  data: any = {}
+}
+
+export namespace DataObject {
+  /**
+   * Creates a proxy that marks the object as dirty whenever this property is modified
+   */
+  export function field(target: AbstractDataObject, propertyKey: string) {
+    if (delete (target as any)[propertyKey]) {
```
[the remark this answers is not in the document]
Comment (belohlavek) [reply src/core/DataObject.ts]: It deletes the original property and replaces it with the proxy (see defineProperty below).
What do you think?

File: src/core/DataObject.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    }
+  }
+
+  export function setValue(target: AbstractDataObject, key: string, value: any) {
+    if (!this.data) {
+      // hack for property initializers
+      this.data = {}
+    }
+
+    if (typeof value === 'object') {
```
[the remark this answers is not in the document]
Comment (belohlavek) [reply src/core/DataObject.ts]: jejeje.. fuck

File: src/core/Engine.ts [hunk 3]
```diff
@@ -61,34 +72,40 @@ export class Engine {
...
-
-      if (entity.hasComponent('transform')) {
-        const t = entity.getComponent<Transform>('transform')
-        obj.position = t.position
-        obj.rotation = t.rotation
-        obj.scale = t.scale
-      }
+      for (let component of Object.values(entity.components)) {
+        if (component.isDirty) {
+          this.componentHandlers[component.type].call(this, entity)
```
[the remark this answers is not in the document]
Comment (belohlavek) [reply src/core/Engine.ts]: yup, super missed it

File: src/core/Component.ts [hunk 4]
```diff
@@ -1,14 +1,28 @@
...
+  readonly type: ComponentType
+
+  private _type: ComponentType
 
   constructor(type: ComponentType) {
-    this.type = type
+    super()
+    this._type = type
+
+    Object.defineProperty(this, 'type', {
```
[the remark this answers is not in the document]
Comment (belohlavek) [reply src/core/Component.ts]: Dang, I thought I couldn't initialize readonlys in a constructor

File: src/core/DataObject.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (belohlavek) [reply src/core/DataObject.ts]: Removed readonlys, they create proxy instances (we should avoid those as much as we can) and this is not decentraland where we care about userland scripts :)

File: src/core/DataObject.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (belohlavek) [reply src/core/DataObject.ts]: Explicit check for arrays and objects

File: src/core/Engine.ts [hunk 5]
```diff
@@ -21,6 +28,10 @@ export class Engine {
 
   private constructor(app: PIXI.Application) {
     this.app = app
+
+    if (process.env.NODE_ENV === 'dev') {
+      window['engine'] = this
+    }
```
[the remark this answers is not in the document]
Comment (belohlavek) [reply src/core/Engine.ts]: moved to index

**Comment by NicoSantangelo** [comment]
🎉

**Pull request merged**